subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/FileChange.java

```diff
@@ -18,22 +18,55 @@
 
 import com.google.common.base.Objects;
 import org.gradle.api.tasks.incremental.InputFileDetails;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
 
 import java.io.File;
 
 public class FileChange implements TaskStateChange, InputFileDetails {
     private final String path;
     private final ChangeType change;
     private final String fileType;
+    private final FileType previousType;
+    private final FileType currentType;
 
-    public FileChange(String path, ChangeType change, String fileType) {
+    public static FileChange added(String path, String title) {
+        return new FileChange(path, ChangeType.ADDED, title, null, null);
+    }
+
+    public static FileChange removed(String path, String title) {
+        return new FileChange(path, ChangeType.REMOVED, title, null, null);
+    }
+
+    public static FileChange modified(String path, String title, FileType previousType, FileType currentType) {
+        assert previousType != null;
+        assert currentType != null;
+        assert !(previousType == FileType.Missing && currentType == FileType.Missing);
+        assert !(previousType == FileType.Directory && currentType == FileType.Directory);
```

Comment:
- [ ] I'm a bit wary about using `assert`. If these are preconditions we want to check, we should use `Preconditions.checkNotNull()` etc.